File: java/org.kevoree.library.java.javaNode/src/main/java/org/kevoree/library/java/wrapper/WrapperFactory.java
package org.kevoree.library.java.wrapper;

import org.kevoree.*;
import org.kevoree.api.BootstrapService;
import org.kevoree.api.ModelService;
import org.kevoree.pmodeling.api.KMFContainer;

/**
 * Created by duke on 9/26/14.
 */
public class WrapperFactory {

    public WrapperFactory(String nodeName) {
        this.nodeName = nodeName;
    }

    private String nodeName;

    public KInstanceWrapper wrap(KMFContainer modelElement, Object newBeanInstance, ThreadGroup tg, BootstrapService bs, ModelService modelService) throws Exception {

        KInstanceWrapper wrapper = null;
        if (modelElement instanceof ComponentInstance) {
            wrapper = new ComponentWrapper();
        }
        if (modelElement instanceof Group) {
            wrapper = new GroupWrapper();
        }
        if (modelElement instanceof Channel) {
            wrapper = new ChannelWrapper();
        }
        if (modelElement instanceof ContainerNode) {
            wrapper = new NodeWrapper();
        }
        if (wrapper != null) {
            wrapper.setModelService(modelService);
            wrapper.setBs(bs);
            wrapper.setIsStarted(false);
            wrapper.setTg(tg);
            wrapper.setNodeName(nodeName);
            wrapper.setTargetObj(newBeanInstance);
            wrapper.setModelElement((Instance) modelElement);
            return wrapper;
        } else {
            throw new Exception("Unknow instance type " + modelElement.metaClassName());
        }
    }

}

File: java/org.kevoree.library.java.javaNode/src/main/java/org/kevoree/library/java/wrapper/WrapperFactory.kt
package org.kevoree.library.java.wrapper

import org.kevoree.modeling.api.KMFContainer
import org.kevoree.ComponentInstance
import org.kevoree.Group
import org.kevoree.Channel
import org.kevoree.ContainerNode
import org.kevoree.api.BootstrapService
import org.kevoree.api.ModelService

/**
 * Created with IntelliJ IDEA.
 * User: duke
 * Date: 03/12/2013
 * Time: 09:33
 */

open class WrapperFactory(val nodeName: String) {

    open fun wrap(modelElement: KMFContainer, newBeanInstance: Any, tg: ThreadGroup, bs: BootstrapService, modelService : ModelService): KInstanceWrapper {
        when(modelElement) {
            is ComponentInstance -> {
                return ComponentWrapper(modelElement, newBeanInstance, nodeName, tg, bs)
            }
            is Group -> {
                return GroupWrapper(modelElement, newBeanInstance, nodeName, tg, bs)
            }
            is Channel -> {
                return ChannelWrapper(modelElement, newBeanInstance, nodeName, tg, bs, modelService)
            }
            is ContainerNode -> {
                return NodeWrapper(modelElement, newBeanInstance, nodeName, tg, bs)
            }
            else -> {
                throw Exception("Unknow instance type " + modelElement.metaClassName())
            }
        }
    }

}
